src/on-stored.cpp

```diff
@@ -0,0 +1,25 @@
+#include <core.h>
+#include <dicom-checksum.h>
+
+OrthancPluginErrorCode OnStoredInstanceCallback(const OrthancPluginDicomInstance *instance, const char *instanceId) {
+    char msg[1024] = {0};
+    sprintf(msg, "OnStored Instance Id: %s", instanceId);
+    //update_checksum(instanceId);
+    DEBUG_LOG(0,msg);
+    DicomChecksum::update_checksum(instanceId);
```

Comment:
Looking at how things connect, I wonder if it wouldn't be better to turn this function into a job processor.

So, what if we create a job when we have the `uuid` and can therefor give the job the `uuid`..
With the `DicomFile` refactor present in #106 we can also give it a copy of the `DicomFile` which will keep the buffer allocation alive..
Then on the OnStored callback we could execute the job. By refactoring the job queue.

We wouldn't have a way of knowing 100% that the job corresponds to the dicom instance though.

I'm just trying to think of ways to reduce the number DB queries.

Reply:
That sounds like a good idea, I'm sold. We should do that.

Reply:
Well then, maybe we just refactor the job queue into an unordered map ~queue. If we can map against the instance somehow..

Or maybe we can put a delay on the existing job queue and we don't turn the OnStored into a processor of jobs. A delay is probably the safest course of action, but it would need to be an intelligent delay that doesn't hold things up. Like somehow we devise a way to delay each job by 5 seconds or something, and if the job has waited at least 5 seconds we execute it.

So perhaps we add jobs with timestamps and a wait time. Not sure how to organize that data. Perhaps with `map<timestamp, tuple<delay,job>>`? If we could also order the jobs by delay that'd be great but `map<TS, map<delay, job>>` seems excessive

Reply:
Now looking at it, the only way to reduce a db call is to do the delay, since no matter what we'll only have either the uuid or the instanceid, but not both at the same time. I don't know how reliable a time delay will be tho.

Reply:
Honestly, I can probably merge the two db call I have now into one, just have to fiddle with sql a bit

Reply:
Maybe what we can do is match based on the file itself. So the idea would be that we can parse pretty quickly (not idea how this compares to DB queries), so we could save the filtered file.. or unfiltered I suppose..(?) then on the other end we parse the content it thinks it stored and compare. We could even just use md5 hashes to compare.

So now it would look like..
**save to disk stack:** create a job that hashes whichever file is important, and saves the hash into a lookup table.. maybe with the uuid
**on stored stack:** create a job that hashes the content, finds the matching entry in the lookup table and update the DB
What do you think?

Reply:
Perhaps to save on work the md5 has to do we tell it to hash a subset of the content. Like maybe only from the first `0010` group element to 2000 elements later

Reply:
Indeed, this works:
```sql
UPDATE attachedfiles
SET uncompressedsize = $1, compressedsize = $2, uncompressedhash = $3, compressedhash = $4 
WHERE uuid = (
  SELECT a.uuid 
  FROM attachedfiles a, resources r 
  WHERE r.internalid = a.id AND r.publicid = $5
)
```